README.md
# Ansible Module for GNS3
[Ansible-Galaxy collections](https://galaxy.ansible.com/davidban77/gns3) repository for GNS3 Server REST API using [gns3fy - see the docs](https://davidban77.github.io/gns3fy/).

## Installation

For the module to be used you need to have installed [gns3fy](https://github.com/davidban77/gns3fy)

```
pip install gns3fy
```

This collections is packaged under ansible-galaxy, so to install it you need [mazer from Ansible Projects](https://galaxy.ansible.com/docs/mazer/index.html):

```
mazer install davidban77.gns3
```

## Features

- Open/closes projects.
- Starts/stops all nodes inside a project, or it can be done sequentially with a delay factor.
- Creates/Updates projects with nodes and links specified as variables in a playbook.
- Deletes projects safely by stopping nodes, if there are any, then closing the project and finally deleting it.
- Creates/Deletes/Restores snapshots of projects.
- Retrieves information about available emulators on the GNS3 server compute, as well as available images, console ports, version, etc..
- Idempotency is present in all actions. An example could be reflected in a playbook that creates a project with nodes and links, these settings will not be executed again on a rerun (and by settings I mean projects settings, nodes and links)/


## Modules

These are the modules provided with this collection:

- `gns3_version`: Retrieves GNS3 server version. (**TO BE DEPRECATED** with the `gns3_facts` module)
- `gns3_facts`: Retrieves the compute(s) information of a GNS3 server
- `gns3_project`: Module to interact with GNS3 server projects
    - It opens/closes projects and performs basic turnup/teradown operations on nodes.
    - It creates/updates or deletes projects, with the respective nodes and links specified
- `gns3_project_file`: Updates/creates a file on a project directory.
- `gns3_snapshot`: Module that interacts with snapshots of a project on GNS3 server.
- `gns3_node`: Module to operate a node in a GNS3 server project.
- `gns3_node_file`: Updates/creates a file on a node directory.
- `gns3_nodes_inventory`: Retrieves GNS3 a project nodes console information.

## Examples: using the module

Here are some examples of how to use the module.

#### Get server version

```yaml
---
- host: localhost
  # Call the collections to use the respective modules
  collections:
    - davidban77.gns3
  vars:
    gns3_url: http://localhost
  tasks:
    - name: Get the server facts
      gns3_facts:
        url: "{{ gns3_url }}"
        port: 3080
        get_images: all
        get_compute_ports: yes
      register: result

    - debug: var=result
```
#### Get nodes_inventory information from a project

```yaml
---
- host: localhost
  # Call the collections to use the respective modules
  collections:
    - davidban77.gns3
  vars:
    gns3_url: http://localhost
  tasks:
    - name: Get the server version
      gns3_nodes_inventory:
        url: "{{ gns3_url }}"
        project_name: lab_example
      register: result

    - debug: var=result
```

#### Manipulate GNS3 projects

```yaml
---
# Open a GNS3 project
- name: Start lab
  gns3_project:
    url: "{{ gns3_url }}"
    state: opened
    project_name: lab_example

# Stop all nodes inside an open project
- name: Stop nodes
  gns3_project:
    url: "{{ gns3_url }}"
    state: opened
    project_name: lab_example
    nodes_state: stopped
    nodes_strategy: all
    poll_wait_time: 5

# Open a GNS3 project and start nodes one by one with a delay of 10sec between them
- name: Start nodes one by one
  gns3_project:
    url: "{{ gns3_url }}"
    state: opened
    project_name: lab_example
    nodes_state: started
    nodes_strategy: one_by_one
    nodes_delay: 10

# Close a GNS3 project
- name: Stop lab
  gns3_project:
    url: "{{ gns3_url }}"
    state: closed
    project_id: "UUID-SOMETHING-1234567"
```

#### Create and delete projects

```yaml
---
# Create a GNS3 project given nodes and links specifications
- name: Create a project
  gns3_project:
    url: "{{ gns3_url }}"
    state: present
    project_name: new_lab
    nodes_spec:
        - name: alpine-1
          template: alpine
        - name: alpine-2
          template: alpine
    links_spec:
        - ['alpine-1', 'eth0', 'alpine-2', 'eth1']

# Delete a GNS3 project
- name: Delete project
  gns3_project:
    url: "{{ gns3_url }}"
    state: absent
    project_name: new_lab
```

## Examples: using the roles

There are also some convinient roles that you can use to manage your labs. Here is an example playbook:

`main.yml`
```yaml
- hosts: localhost
  tasks:
    - import_role:
        name: create_lab
      when: execute == "create"
    - import_role:
        name: delete_lab
      when: execute == "delete"
```

This way you can call and switch the behaviour of the playbook:

**Create the lab**
```
ansible-playbook main.yml -e execute=create
```

Or **delete the lab**
```
ansible-playbook main.yml -e execute=delete
```

Here is the example variable file which specifies the naming convention used. You can see that the variable names come from the module itself with only `gns3_`

```yaml
---
gns3_url: "http://dev_gns3server"
gns3_project_name: test_ansible
gns3_nodes_spec:
    - name: veos-1
      template: "vEOS-4.21.5F"
    - name: veos-2
      template: "vEOS-4.21.5F"
    - name: ios-1
      template: "IOU-15.4"
    - name: ios-2
      template: "IOU-15.4"
gns3_nodes_strategy: one_by_one
gns3_links_spec:
    - ["veos-1", "Ethernet1", "veos-2", "Ethernet1"]
    - ["veos-1", "Ethernet2", "ios-1", "Ethernet1/0"]
    - ["veos-2", "Ethernet2", "ios-2", "Ethernet1/0"]
    - ["ios-1", "Ethernet1/2", "ios-2", "Ethernet1/2"]
```

### More examples

For more examples like create an `/etc/network/interfaces` file for an `alpine` docker node to configure its network interfaces, or restore a project to an specific snapshot, you can go to the `test/playbooks` directory.


CHANGELOG.md
# Releases

## 1.5.0

**Enhancements:**

- Added `gns3_facts` to retrieve the compute information like: console ports, server version, available emulators, available images, etc..
- Added `gns3_snapshot` to manipulate the snapshot creation/deletion and restoration of a project.

**Fixes:**

- Added the needed `user` and `password` arguments to all the modules, when interacting with a GNS3 server with authentication.
- Refactored the modules a little to be more standard with each other.

## 1.4.0

**Enhacements:**

- Added `gns3_node` to interact with the node inside a project. Provides the following:
    - `start/stop/suspend/reload`: Actions to be applied on the node. These are idempotent with the exception of `reload`.
    - Special flags like `retry` for an action to be applied a second time just in case... And a `force_project_open` to interact with a device if the project is closed
- Refactored `gns3_project` to be more pythonic

## 1.3.0

**Enhancements:**

- Added `gns3_node_file` and `gns3_project_file` modules.
- Improved the Makefile
- Added alpine node to the tests

## 1.2.2.

**Fixes:**

- Upgrading to `gns3fy ^0.4.0`

## 1.2.1

**Enhancements:**

- No more `node_type` needed when creating nodes in a project.

## 1.2.0

**Enhancements:**

- Modules:
    - `gns3_nodes_inventory`: Returns inventory-style dictionary of the nodes.

**Fixes:**

- Modules:
    - Error when using the `gns3_version` module when `gns3fy` is not installed

- Tests:
    - Added check for `gns3_version`

## 1.1.0

**Enhancements:**

- Roles:
    - `create_lab`: Create a GNS3 Lab by creating a project and starting up the nodes
    - `delete_lab`: Deletes a GNS3 Lab by stopping nodes and deleting the project

## 1.0.1

Initial push of the ansible collections. Released focused on the `gns3` module


plugins/modules/gns3_node.py
#!/usr/bin/env python

ANSIBLE_METADATA = {
    "metadata_version": "1.2",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: gns3_node
short_description: Module to operate a node in a GNS3 server project
version_added: '2.8'
description:
    - Module to operate a node in a GNS3 server project.
    - It starts/stops/suspend/reloads a node.
requirements: [ gns3fy ]
author:
    - David Flores (@davidban77)
options:
    url:
        description:
            - URL target of the GNS3 server
        required: true
        type: str
    port:
        description:
            - TCP port to connect to server REST API
        type: int
        default: 3080
    user:
        description:
            - User to connect to GNS3 server
        type: str
    password:
        description:
            - Password to connect to GNS3 server
        type: str
    project_name:
        description:
            - Project name
        type: str
    project_id:
        description:
            - Project ID
        type: str
    node_name:
        description:
            - Node name
        type: str
    node_id:
        description:
            - Node ID
        type: str
    state:
        description:
            - State of the node, it can be:
            - '- C(started): Starts a node'
            - '- C(stopped): Stops a node'
            - '- C(suspended): Suspends a node'
            - '- C(reload): Special non-idempotent action that reloads a node'
        type: str
        choices: ['started', 'stopped', 'suspended', 'reload']
    retry:
        description:
            - Retries an action based on the state, if true and state is set to reload
            - it will reload the device and try to start it again if the status was not
            - changed
        type: bool
        default: false
    poll_wait_time:
        description:
            - Delay in seconds to wait to poll nodes when they are started/stopped.
            - Used when I(nodes_state) is C(started)/C(stopped)
        type: int
        default: 5
    force_project_open:
        description:
            - It will open the project (if closed) to interact with the device.
            - Otherwise it will throw out an error
        type: bool
        default: false
"""

EXAMPLES = """
# Open a GNS3 project and start router01 node
- name: Start node
  gns3_node:
    url: http://localhost
    project_name: lab_example
    node_name: router01
    node_state: started
    force_project_open: true

# Stop a node and wait 10 seconds to poll for status
- name: Stop node
  gns3_node:
    url: http://localhost
    project_name: lab_example
    node_name: router01
    state: stopped

# Suspend a node based on UUID
- name: Suspend node
  gns3_node:
    url: http://localhost
    project_name: lab_example
    node_id: 'ROUTER-UUID-SOMETHING-1234567'
    state: suspended

# Reload a node and apply a retry to start if needed
- name: Stop lab
  gns3_node:
    url: http://localhost
    project_id: 'PROJECT-UUID-SOMETHING-1234567'
    node_name: router01
    state: reload
    retry: true
    poll_wait_time: 30
"""

RETURN = """
name:
    description: Project name
    type: str
project_id:
    description: Project UUID
    type: str
node_id:
    description: Node UUID
    type: str
status:
    description: Project status. Possible values: opened, closed
    type: str
node_directory:
    description: Path of the node on the server (works only with compute=local)
    type: str
node_type:
    description: Network node type
    type: str
"""
import time
import traceback
from ansible.module_utils.basic import AnsibleModule, missing_required_lib

GNS3FY_IMP_ERR = None
try:
    from gns3fy import Gns3Connector, Project

    HAS_GNS3FY = True
except Exception:
    HAS_GNS3FY = False
    GNS3FY_IMP_ERR = traceback.format_exc()


def return_node_data(node):
    "Returns the node main attributes"
    return dict(
        name=node.name,
        project_id=node.project_id,
        node_id=node.node_id,
        status=node.status,
        node_directory=node.node_directory,
        node_type=node.node_type,
    )


def state_verification(expected_state, node, retry=False, poll_wait_time=5):
    "Verifies node state and returns a changed attribute"
    if expected_state == "started" and node.status != "started":
        node.start()
        if node.status != "started" and retry:
            node.start()
        return True
    elif expected_state == "stopped" and node.status != "stopped":
        node.stop()
        if node.status != "stopped" and retry:
            node.stop()
        return True
    elif expected_state == "suspended" and node.status != "suspended":
        node.suspend()
        if node.status != "suspended" and retry:
            node.suspend()
        return True
    elif expected_state == "reload":
        node.reload()
        time.sleep(poll_wait_time)
        node.get()
        if node.status != "started" and retry:
            node.start()
        return True
    return False


def main():
    module = AnsibleModule(
        argument_spec=dict(
            url=dict(type="str", required=True),
            port=dict(type="int", default=3080),
            user=dict(type="str", default=None),
            password=dict(type="str", default=None, no_log=True),
            state=dict(
                type="str",
                required=True,
                choices=["started", "stopped", "suspended", "reload"],
            ),
            project_name=dict(type="str", default=None),
            project_id=dict(type="str", default=None),
            node_name=dict(type="str", default=None),
            node_id=dict(type="str", default=None),
            retry=dict(type="bool", default=False),
            poll_wait_time=dict(type="int", default=5),
            force_project_open=dict(type="bool", default=False),
        ),
        required_one_of=[["project_name", "project_id"], ["node_name", "node_id"]],
    )
    result = dict(changed=False)
    if not HAS_GNS3FY:
        module.fail_json(msg=missing_required_lib("gns3fy"), exception=GNS3FY_IMP_ERR)

    server_url = module.params["url"]
    server_port = module.params["port"]
    server_user = module.params["user"]
    server_password = module.params["password"]
    state = module.params["state"]
    project_name = module.params["project_name"]
    project_id = module.params["project_id"]
    node_name = module.params["node_name"]
    node_id = module.params["node_id"]
    retry = module.params["retry"]
    poll_wait_time = module.params["poll_wait_time"]
    force_project_open = module.params["force_project_open"]

    try:
        # Create server session
        server = Gns3Connector(
            url=f"{server_url}:{server_port}", user=server_user, cred=server_password
        )
        # Define the project
        if project_name is not None:
            project = Project(name=project_name, connector=server)
        elif project_id is not None:
            project = Project(project_id=project_id, connector=server)
        if project is None:
            module.fail_json(msg="Could not retrieve project. Check name", **result)

        project.get()
        if project.status != "opened" and force_project_open:
            project.open()

        # Retrieve node
        if node_name is not None:
            node = project.get_node(name=node_name)
        elif node_id is not None:
            node = project.get_node(node_id=node_id)
        if node is None:
            module.fail_json(msg="Could not retrieve node. Check name", **result)
    except Exception as err:
        module.fail_json(msg=str(err), **result)

    # Apply state change
    result["changed"] = state_verification(
        expected_state=state, node=node, retry=retry, poll_wait_time=poll_wait_time
    )

    # Return the node data
    result["node"] = return_node_data(node)
    module.exit_json(**result)


if __name__ == "__main__":
    main()


plugins/modules/gns3_facts.py
#!/usr/bin/env python

ANSIBLE_METADATA = {
    "metadata_version": "1.1",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: gns3_facts
short_description: Module that retrieves the compute(s) information of a GNS3 server
version_added: '2.8'
description:
    - Module that retrieves the compute(s) information of a GNS3 server
requirements: [ gns3fy ]
author:
    - David Flores (@davidban77)
options:
    url:
        description:
            - URL target of the GNS3 server
        required: true
        type: str
    port:
        description:
            - TCP port to connect to server REST API
        type: int
        default: 3080
    user:
        description:
            - User to connect to GNS3 server
        type: str
    password:
        description:
            - Password to connect to GNS3 server
        type: str
    get_images:
        description:
            - If set it will also retrieve the images of the specified emulator unless
            - is set to 'all', in which case will retrieve from all emulators. For a
            - list of available emulators, visit the GNS3 API information
        type: str
    get_compute_ports:
        description:
            - If set it will retrieve the console_ports and udp_ports of the compute
        type: bool
"""

EXAMPLES = """
# Retrieves all the  information from the computes of GNS3 server
- name: Retrieve all the facts of a GNS3 server computes
  gns3_facts:
    url: http://localhost
    get_images: all
    get_compute_ports: yes
  register: computes_info

- debug: var=computes_info

# Retrieves only basic facts data of the GNS3 server computes
- gns3_facts:
    url: http://localhost
  register: computes_info

- debug: var=computes_info
"""

RETURN = """
compute_id:
    description: Server identifier
    type: str
name:
    description: Server name
    type: str
host:
    description: Server host
    type: str
capabilities:
    description: Object that describes what the server supports
    type: dict
connected:
    description: Whether the controller is connected to the compute or not
    type: bool
cpu_usage_percent:
    description: CPU usage of the compute
    type: float
memory_usage_percent:
    description: RAM usage of the compute
    type: int
port:
    description: Server port
    type: int
protocol:
    description: Protocol used (http, https)
    type: str
user:
    description: User for authentication
    type: str
last_error:
    description: Last error on the compute
    type: str
images:
    description: Images configured on the compute depending on the emulator (optional)
    type: dict
compute_ports:
    description: Ports used by the compute (console and udp ports) (optional)
    type: dict
"""

import traceback
from ansible.module_utils.basic import AnsibleModule, missing_required_lib

GNS3FY_IMP_ERR = None
try:
    from gns3fy import Gns3Connector

    HAS_GNS3FY = True
except Exception:
    HAS_GNS3FY = False
    GNS3FY_IMP_ERR = traceback.format_exc()


def main():
    module = AnsibleModule(
        argument_spec=dict(
            url=dict(type="str", required=True),
            port=dict(type="int", default=3080),
            user=dict(type="str", default=None),
            password=dict(type="str", default=None, no_log=True),
            get_images=dict(type="str", default=None),
            get_compute_ports=dict(type="bool", default=False),
        )
    )
    result = dict(changed=False)
    if not HAS_GNS3FY:
        module.fail_json(msg=missing_required_lib("gns3fy"), exception=GNS3FY_IMP_ERR)

    server_url = module.params["url"]
    server_port = module.params["port"]
    server_user = module.params["user"]
    server_password = module.params["password"]
    get_images = module.params["get_images"]
    get_compute_ports = module.params["get_compute_ports"]

    try:
        # Create server session
        server = Gns3Connector(
            url=f"{server_url}:{server_port}", user=server_user, cred=server_password
        )

        computes = server.get_computes()
        for compute in computes:

            # Images
            if get_images:
                compute["images"] = dict()

                if get_images == "all":
                    for emulator in compute["capabilities"]["node_types"]:
                        try:
                            compute["images"][emulator] = server.get_compute_images(
                                emulator=emulator, compute_id=compute["compute_id"]
                            )
                        except Exception as err:
                            if "404" in str(err):
                                # Contine if no image dir is set for that emulator
                                continue
                else:
                    compute["images"][get_images] = server.get_compute_images(
                        emulator=get_images, compute_id=compute["compute_id"]
                    )

            # Compute ports
            if get_compute_ports:
                compute["compute_ports"] = server.get_compute_ports(
                    compute_id=compute["compute_id"]
                )

        result["facts"] = computes
        module.exit_json(**result)
    except Exception as err:
        module.fail_json(msg=str(err), **result)


if __name__ == "__main__":
    main()


plugins/modules/gns3_project.py
#!/usr/bin/env python

ANSIBLE_METADATA = {
    "metadata_version": "1.4",
    "status": ["preview"],
    "supported_by": "community",
}

DOCUMENTATION = """
---
module: gns3_project
short_description: Module to interact with GNS3 server projects
version_added: '2.8'
description:
    - 'Module to interact with GNS3 server projects.
    - It is using the L(gns3fy library,https://davidban77.github.io/gns3fy/)'
    - It opens/closes projects and performs basic turnup/teradown operations on nodes.
    - It creates/updates or deletes projects.
requirements: [ gns3fy ]
author:
    - David Flores (@davidban77)
options:
    url:
        description:
            - URL target of the GNS3 server
        required: true
        type: str
    port:
        description:
            - TCP port to connect to server REST API
        type: int
        default: 3080
    user:
        description:
            - User to connect to GNS3 server
        type: str
    password:
        description:
            - Password to connect to GNS3 server
        type: str
    state:
        description:
            - State of the project to be on the GNS3 server
            - '- C(opened): Opens a project and turns up nodes'
            - '- C(closed): Closes a project and turns down nodes'
            - '- C(present): Creates/update a project on the server'
            - '- C(absent): Deletes a project on the server'
        type: str
        choices: ['opened', 'closed', 'present', 'absent']
    project_name:
        description:
            - Project name
        type: str
    project_id:
        description:
            - Project ID
        type: str
    nodes_state:
        description:
            - Starts/stops nodes on the project.
            - Used when I(state) is C(opened)/C(closed)
        type: str
        choices: ['started', 'stopped']
    nodes_strategy:
        description:
            - Start/stop strategy of the devices defined on the project.
            - '- C(all): It starts/stops all nodes at once'
            - '- C(one_by_one): It starts/stops nodes serialy using I(nodes_delay) time
            between each action'
            - Used when I(state) is C(opened)/C(closed)
        type: str
        choices: ['all', 'one_by_one']
        default: 'all'
    nodes_delay:
        description:
            - Delay time in seconds to wait between nodes start/stop
            - Used when I(nodes_strategy) is C(one_by_one)
        type: int
        default: 10
    poll_wait_time:
        description:
            - Delay in seconds to wait to poll nodes when they are started/stopped.
            - Used when I(nodes_state) is C(started)/C(stopped)
        type: int
        default: 5
    nodes_spec:
        description:
            - List of dictionaries specifying the nodes properties.
            - '- Mandatory attributes: C(name), C(node_type) and C(template).'
            - '- Optional attributes: C(compute_id). It defaults to C(local)'
        type: list
    links_spec:
        description:
            - 'List of lists specifying the links endpoints. Example: C(- ["alpine-1",
            "eth0", "alpine-2", "eth0"])'
            - 'Mandatory attributes: C(node_a), C(port_a), C(node_b) and C(port_b)'
        type: list
"""

EXAMPLES = """
# Open a GNS3 project
- name: Start lab
  gns3_project:
    url: http://localhost
    state: opened
    project_name: lab_example

# Stop all nodes inside an open project
- name: Stop nodes
  gns3_project:
    url: http://localhost
    state: opened
    project_name: lab_example
    nodes_state: stopped
    nodes_strategy: all
    poll_wait_time: 5

# Open a GNS3 project and start nodes one by one with a delay of 10sec between them
- name: Start nodes one by one
  gns3_project:
    url: http://localhost
    state: opened
    project_name: lab_example
    nodes_state: started
    nodes_strategy: one_by_one
    nodes_delay: 10

# Close a GNS3 project
- name: Stop lab
  gns3_project:
    url: http://localhost
    state: closed
    project_id: 'UUID-SOMETHING-1234567'

# Create a GNS3 project
- name: Create a project given nodes and links specs
  gns3_project:
    url: http://localhost
    state: present
    project_name: new_lab
    nodes_spec:
        - name: alpine-1
          node_type: docker
          template: alpine
        - name: alpine-2
          node_type: docker
          template: alpine
    links_spec:
        - ('alpine-1', 'eth0', 'alpine-2', 'eth1')

# Delete a GNS3 project
- name: Delete project
  gns3_project:
    url: http://localhost
    state: absent
    project_name: new_lab
"""

RETURN = """
name:
    description: Project name
    type: str
project_id:
    description: Project UUID
    type: str
status:
    description: Project status. Possible values: opened, closed
    type: str
path:
    description: Path of the project on the server (works only with compute=local)
    type: str
auto_close:
    description: Project auto close when client cut off the notifications feed
    type: bool
auto_open:
    description: Project open when GNS3 start
    type: bool
auto_start:
    description: Project start when opened
    type: bool
filename:
    description: Project filename
    type: str
"""

import time
import traceback
from ansible.module_utils.basic import AnsibleModule, missing_required_lib

GNS3FY_IMP_ERR = None
try:
    from gns3fy import Gns3Connector, Project

    HAS_GNS3FY = True
except Exception:
    HAS_GNS3FY = False
    GNS3FY_IMP_ERR = traceback.format_exc()


def return_project_data(project):
    "Returns the project main attributes"
    return dict(
        name=project.name,
        project_id=project.project_id,
        status=project.status,
        path=project.path,
        auto_close=project.auto_close,
        auto_open=project.auto_open,
        auto_start=project.auto_start,
        filename=project.filename,
    )


# def nodes_state_verification(module, project, result):
def nodes_state_verification(
    expected_nodes_state, nodes_strategy, nodes_delay, poll_wait_time, project
):
    "Verifies each node state and returns a changed attribute"
    nodes_statuses = [node.status for node in project.nodes]

    # Verify if nodes do not match expected state
    if expected_nodes_state == "started" and any(
        status == "stopped" for status in nodes_statuses
    ):
        # Turnup the nodes based on strategy
        if nodes_strategy == "all":
            project.start_nodes(poll_wait_time=poll_wait_time)
        elif nodes_strategy == "one_by_one":
            for node in project.nodes:
                if node.status != "started":
                    node.start()
                    time.sleep(nodes_delay)
        return True
    elif expected_nodes_state == "stopped" and any(
        status == "started" for status in nodes_statuses
    ):
        # Shutdown nodes based on strategy
        if nodes_strategy == "all":
            project.stop_nodes(poll_wait_time=poll_wait_time)
        elif nodes_strategy == "one_by_one":
            for node in project.nodes:
                if node.status != "stopped":
                    node.stop()
                    time.sleep(nodes_delay)
        return True
    return False


def create_node(node_spec, project, module):
    "Creates the node specified in nodes_spec"
    # If exceptions occur then print them out in ansible format
    try:
        project.create_node(**node_spec)
    except Exception as err:
        module.fail_json(msg=str(err), exception=traceback.format_exc())


def create_link(link_spec, project, module):
    "Creates the node specified in nodes_spec"
    # If exceptions occur then print them out in ansible format
    try:
        project.create_link(*link_spec)
    except ValueError as err:
        if "At least one port is used" in str(err):
            return False
        else:
            module.fail_json(msg=str(err), exception=traceback.format_exc())
    except Exception as err:
        module.fail_json(msg=str(err), exception=traceback.format_exc())
    return True


def main():
    module = AnsibleModule(
        argument_spec=dict(
            url=dict(type="str", required=True),
            port=dict(type="int", default=3080),
            user=dict(type="str", default=None),
            password=dict(type="str", default=None, no_log=True),
            state=dict(
                type="str",
                required=True,
                choices=["opened", "closed", "present", "absent"],
            ),
            project_name=dict(type="str", default=None),
            project_id=dict(type="str", default=None),
            nodes_state=dict(type="str", choices=["started", "stopped"]),
            nodes_strategy=dict(
                type="str", choices=["all", "one_by_one"], default="all"
            ),
            nodes_delay=dict(type="int", default=10),
            poll_wait_time=dict(type="int", default=5),
            nodes_spec=dict(type="list"),
            links_spec=dict(type="list"),
        ),
        supports_check_mode=True,
        required_one_of=[["project_name", "project_id"]],
        required_if=[["nodes_strategy", "one_by_one", ["nodes_delay"]]],
    )
    result = dict(changed=False)
    if not HAS_GNS3FY:
        module.fail_json(msg=missing_required_lib("gns3fy"), exception=GNS3FY_IMP_ERR)
    if module.check_mode:
        module.exit_json(**result)

    server_url = module.params["url"]
    server_port = module.params["port"]
    server_user = module.params["user"]
    server_password = module.params["password"]
    state = module.params["state"]
    project_name = module.params["project_name"]
    project_id = module.params["project_id"]
    nodes_state = module.params["nodes_state"]
    nodes_strategy = module.params["nodes_strategy"]
    nodes_delay = module.params["nodes_delay"]
    poll_wait_time = module.params["poll_wait_time"]
    nodes_spec = module.params["nodes_spec"]
    links_spec = module.params["links_spec"]

    try:
        # Create server session
        server = Gns3Connector(
            url=f"{server_url}:{server_port}", user=server_user, cred=server_password
        )
        # Define the project
        if project_name is not None:
            project = Project(name=project_name, connector=server)
        elif project_id is not None:
            project = Project(project_id=project_id, connector=server)
    except Exception as err:
        module.fail_json(msg=str(err), **result)

    #  Retrieve project information
    try:
        project.get()
        pr_exists = True
    except Exception as err:
        pr_exists = False
        reason = str(err)

    if state == "opened":
        if pr_exists:
            if project.status != "opened":
                # Open project
                project.open()

                # Now verify nodes
                if nodes_state is not None:

                    # Change flag based on the nodes state
                    result["changed"] = nodes_state_verification(
                        expected_nodes_state=nodes_state,
                        nodes_strategy=nodes_strategy,
                        nodes_delay=nodes_delay,
                        poll_wait_time=poll_wait_time,
                        project=project,
                    )
                else:
                    # Means that nodes are not taken into account for idempotency
                    result["changed"] = True
            # Even if the project is open if nodes_state has been set, check it
            else:
                if nodes_state is not None:
                    result["changed"] = nodes_state_verification(
                        expected_nodes_state=nodes_state,
                        nodes_strategy=nodes_strategy,
                        nodes_delay=nodes_delay,
                        poll_wait_time=poll_wait_time,
                        project=project,
                    )

        else:
            module.fail_json(msg=reason, **result)

    elif state == "closed":
        if pr_exists:
            if project.status != "closed":
                # Close project
                project.close()
                result["changed"] = True
        else:
            module.fail_json(msg=reason, **result)

    elif state == "present":
        if pr_exists:
            if nodes_spec is not None:
                # Need to verify if nodes exist
                _nodes_already_created = [node.name for node in project.nodes]
                for node_spec in nodes_spec:
                    if node_spec["name"] not in _nodes_already_created:
                        # Open the project in case it was closed
                        project.open()
                        create_node(node_spec, project, module)
                        result["changed"] = True
            if links_spec is not None:
                for link_spec in links_spec:
                    project.open()
                    # Trigger another get to refresh nodes attributes
                    project.get()
                    # Link verification is already built in the library
                    created = create_link(link_spec, project, module)
                    if created:
                        result["changed"] = True
        else:
            # Create project
            project.create()
            # Nodes section
            if nodes_spec is not None:
                for node_spec in nodes_spec:
                    create_node(node_spec, project, module)
            # Links section
            if links_spec is not None:
                for link_spec in links_spec:
                    create_link(link_spec, project, module)
            result["changed"] = True
    elif state == "absent":
        if pr_exists:
            # Stop nodes and close project to perform delete gracefully
            if project.status != "opened":
                # Project needs to be opened in order to be deleted...
                project.open()
            project.stop_nodes(poll_wait_time=0)
            project.delete()
            result["changed"] = True
        else:
            module.exit_json(**result)

    # Return the project data
    result["project"] = return_project_data(project)
    module.exit_json(**result)


if __name__ == "__main__":
    main()
